jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/DulmageMendelsohnDecomposition.java

```diff
@@ -0,0 +1,307 @@
+/*
+ * (C) Copyright 2018-2018, by CAE Tech Limited and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.decomposition;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import org.jgrapht.Graph;
+import org.jgrapht.GraphTests;
+import org.jgrapht.alg.connectivity.KosarajuStrongConnectivityInspector;
+import org.jgrapht.alg.interfaces.MatchingAlgorithm;
+import org.jgrapht.alg.interfaces.MatchingAlgorithm.Matching;
+import org.jgrapht.alg.interfaces.StrongConnectivityAlgorithm;
+import org.jgrapht.alg.matching.HopcroftKarpMaximumCardinalityBipartiteMatching;
+import org.jgrapht.graph.DefaultDirectedGraph;
+import org.jgrapht.graph.DefaultEdge;
+import org.jgrapht.graph.EdgeReversedGraph;
+import org.jgrapht.graph.builder.GraphBuilder;
+import org.jgrapht.traverse.DepthFirstIterator;
+
+/**
+ * This class computes a Dulmage-Mendelsohn Decomposition of a bipartite graph. 
+ * A Dulmage–Mendelsohn decomposition is a partition of the vertices of a 
+ * bipartite graph into subsets, with the property that two adjacent vertices 
+ * belong to the same subset if and only if they are paired with each other in a 
+ * perfect matching of the graph. This particular implementation is capable of 
+ * computing both a coarse and a fine Dulmage-Mendelsohn Decomposition.
+ * 
+ * The Dulmage-Mendelsohn Decomposition is based on a maximum-matching of the 
+ * graph G. This implementation uses the Hopcroft-Karp maximum matching 
```

Comment:
Please wrap all math in $$ signs, see: https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation#math-in-javadoc
$G$